Added sql code to empty file

Sailtor opened the pull request:
MOAQ ONLINE

Sailtor committed:
Added sql code to empty file

MOAQ ONLINE

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,434 @@
+
+-- Task 2 (city ID = 4)
+SELECT DISTINCT banks.Name
+FROM banks, branches
```
Think about what ',' does and if it's correct behavior (and in the tasks lower)

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,434 @@
+)
+CREATE TABLE accounts
+(
+	ClientId INT NOT NULL,
```
Better to have a separate column - Id in this concrete case for accounts

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,434 @@
+
+-- Task 4
+SELECT accounts.ClientId, accounts.BankId, accounts.Balance AS AccountBalance, SUM(cards.Balance) AS CardBalance, accounts.Balance - SUM(cards.Balance) AS Mismatch
+FROM accounts
```
You should also show the difference when account doesn't have any cards

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,434 @@
+	JOIN cards on accounts.BankId = cards.BankId 
+		AND accounts.ClientId = cards.ClientId
+GROUP BY accounts.ClientId, accounts.BankId, accounts.Balance
+HAVING accounts.Balance - SUM(cards.Balance) > 0
```
Can cards balance be greater than accounts balance?

kseniya-kalita reviewed (changes requested):
Please update your script in accordance with the review comments and send it back for further review

Sailtor committed:
Update MotherOfAllQueries.sql

Fixes

Sailtor commented:
Updated file according to suggested changes

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+GO
+
+-- Task 1
+-- Creating database tables and constraints for them
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+	(12,800)
+GO
+
+-- Task 2 (city ID = 4)
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+WHERE branches.CityId = 2
+GO
+
+-- Task 3
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+	JOIN banks on accounts.BankId = banks.Id
+GO
+
+-- Task 4
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+GROUP BY socialGroups.Id, socialGroups.Name
+
+-- Task 5 (using subquery)
+SELECT socGroupsHigh.Id, socGroupsHigh.Name, (
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+HAVING accounts.Balance - COALESCE(SUM(cards.Balance),0) != 0
+GO
+
+-- Task 5 (using GROUP BY)
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+FROM socialGroups as socGroupsHigh
+
+-- Task 6 
+CREATE PROCEDURE Add10$ToSocGroup 
```
If i run the whole file in ms sql server, then this procedure is not created and i get the next error. Please fix so the whole file can be run one time and all procedures, tables, triggers are creted in one click 
![image](https://github.com/Sailtor/Innowise.SQLTestTask/assets/96731963/2bcb712c-5b08-4d2c-93fc-52d820e14439)

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+	WHERE socialGroups.Id = socGroupsHigh.Id) AS CardsCount
+FROM socialGroups as socGroupsHigh
+
+-- Task 6 
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+GO
+
+-- Task 7
+SELECT accounts.Id, clients.Name, clients.Surname, accounts.Balance - COALESCE(SUM(cards.Balance),0) AS AvailableFunds
```
Accepted

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+							AND cards.AccountId = @AccountId)
+						BEGIN
+							PRINT 'Specified card does not belong to specified account'
+							ROLLBACK
```
Investigate more on transactions and rollbacks, when it's correct to begin

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+						WHERE cards.Id = @CardId
+							AND cards.AccountId = @AccountId)
+						BEGIN
+							PRINT 'Specified card does not belong to specified account'
```
Be careful with too much descriptive errors. It's fine here but if the api will handle it, it could give too much data for the hackers (especially in finance domain). As a test task, you can leave it as it is, but just keep in mind

Sailtor replied:
Understood, I'll keep it in mind for further work

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+INSTEAD OF UPDATE
+AS
+BEGIN
+	UPDATE accounts
```
As a developer, how do i know if anything was updated or not? it's not obvious that when i update entry, they cannot be updated and i will not see anything. Trying to find a good approach for this. The same for the second trigger

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,413 @@
+AS
+BEGIN
+	UPDATE cards
+	SET cards.Balance = COALESCE((
```
overcomplicated, try to make it easier :)

Sailtor replied:
:)

kseniya-kalita reviewed (changes requested):
Good work! Tasks 1-7 are accepted, 8-9 need some rework. Please review the comments and update your script

Sailtor committed:
Yet another MOAQ update

I feel the power of SQL flowing in my veins....
One day I'll transcend human possibilities and become one with the machine.
Glory to Omnissiah!

Sailtor commented:
Updated file according to suggested changes

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,434 @@
+INSTEAD OF UPDATE
+AS
+BEGIN
+	UPDATE accounts
```
Find another approach, not by using print. The trigger should allow updating all the entries or don't allow to update any (i.e. use throwing errors here). Please apply the same in task 8
_As a developer, how do i know if anything was updated or not? it's not obvious that when i update entry, they cannot be updated and i will not see anything. Trying to find a good approach for this. The same for the second trigger_

kseniya-kalita reviewed (changes requested):
Please review the comments and update the code appropriately

Sailtor committed:
It evolves...

The sacred machine allows me to cast the sacred spells on it
Every day the machine spirit heals, gathering power
I feel blessed witnessing this magnificent process of it becoming closer to the ideals of the Machine God

Sailtor commented:
Updated file according to suggested changes

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,407 @@
+GO
+
+--Task 9
+CREATE TRIGGER Accounts_UPDATE
```
![image](https://github.com/Sailtor/Innowise.SQLTestTask/assets/96731963/b56544aa-9f3f-4022-ac0c-ef80d2298df0)

Please review your script and fix the next error. Also there woud be problems if you would try to update other fields (not balance). So better to choose another type of trigger here or to take into considerations all of this

kseniya-kalita replied:
The same applies to the next trigger

kseniya-kalita commented on MotherOfAllQueries.sql:
```diff
@@ -0,0 +1,407 @@
+AS
+BEGIN
+	BEGIN TRY
+	BEGIN TRANSACTION
```
it's better to start the transaction right before the modifying data, just validation can be outside of it

kseniya-kalita reviewed (changes requested):
Good work with raising errors, please review comments and fix them. Hopefully with the next review we will finish. Don't hurry up, take time and think on the comments :)

Sailtor committed:
Immortal Machine

A promised end waits for me. Were my skills enough to reach it? Or will my accomplishments be forgotten? Will my work be crushed under the relentless march of higher prosperity? Or will I finally ascend to something worthy of Omnissiah's blessing? Who knows....
The time has come to find out.
ALL HAIL THE MACHINE GOD!

Sailtor commented:
Updated the code according to suggested changes

P.S. Maybe hurried up)
In any case, it was an exciting journey

kseniya-kalita reviewed (approved):
All tasks are accepted, good work!
